# Hyper parameter testing

In this section of the code we will try to fine tune our model to get better results.

<hr />

Import cleaned and aggreated data from [02_data_cleaning_and_aggregation.ipynb](./02_data_cleaning_and_aggregation.ipynb)

In [2]:
import pandas as pd
from tqdm import tqdm

df=pd.read_json("./datasets/generated/cleaned_aggregated_data_2.json")

In [3]:
df

tmID  year playoff  averageWinRate  averagePoints  averageRebounds  \
0    WAS    10       Y        0.470588     222.360859       101.006193   
1    WAS     9       N        0.529412     229.818750       101.212500   
2    WAS     8       N        0.470588     208.948315        87.485393   
3    WAS     7       Y        0.500000     234.080838        98.093812   
4    WAS     6       N        0.264706     208.569231        96.557265   
..   ...   ...     ...             ...            ...              ...   
105  CHA     5       N        0.562500     221.324885        99.687623   
106  CHA     4       Y        0.562500     206.986888        93.321678   
107  CHA     3       Y        0.250000     195.899778        75.541420   
108  ATL    10       Y        0.500000     187.791463        94.755002   
109  ATL     9       N        0.500000     193.078571        82.600000   

     averageAssists  averageSteals  averageBlocks  averageTurnovers  \
0         40.109001      22.393889      10.068126         42.180842   
1         52.352083      24.168750       9.056250         46.260417   
2         40.890637      23.185019       8.444944         38.629213   
3         57.265469      25.990519      12.104291         47.100798   
4         45.564103      22.707692      11.046154         40.806838   
..              ...            ...            ...               ...   
105       50.847597      24.255431      11.236011         42.470046   
106       47.524913      24.971591      10.012238         43.409528   
107       53.237056      21.937500       8.870192         44.350592   
108       44.626501      19.664295      10.381948         42.261450   
109       40.688095      18.550000       7.564286         41.083333   

     averageFGRatio  averageFTRatio  averageThreeRatio  coachWinRate  \
0          0.409265        0.766500           0.253256      0.500000   
1          0.430233        0.775180           0.326570      0.500000   
2          0.421434        0.746699           0.342584      0.470588   
3          0.423451        0.748017           0.347804      0.437500   
4          0.398834        0.694540           0.337006      0.470588   
..              ...             ...                ...           ...   
105        0.423319        0.749351           0.323474      0.500000   
106        0.414555        0.791145           0.434595      0.500000   
107        0.407616        0.758836           0.287525      0.281250   
108        0.434338        0.708438           0.393790      0.500000   
109        0.436712        0.706352           0.315101      0.500000   

     numberOfAwardedPlayers  
0                         0  
1                         2  
2                         2  
3                         2  
4                         1  
..                      ...  
105                       0  
106                       0  
107                       0  
108                       5  
109                       2  

[110 rows x 15 columns]

## Logistic regression

### Testing some attributes with SVM

In [13]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split # Import train_test_split function

# Assuming X_train, y_train, X_test, y_test are your training and testing data
# X_train and X_test should be your feature matrices, and y_train and y_test should be your target labels (0 or 1)
features=['averagePoints','averageRebounds','averageAssists','averageFGRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']
target = 'playoff'
x = df[features]
y = df[target]

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=1) # 70% training and 30% test

# Logistic Regression
logreg = LogisticRegression()

logreg = LogisticRegression(max_iter=1000)
logreg.fit(X_train, y_train)
logreg_predictions = logreg.predict(X_test)
logreg_accuracy = accuracy_score(y_test, logreg_predictions)
print("Logistic Regression Accuracy:", logreg_accuracy)

# Support Vector Machine
svm = SVC(kernel='linear')  # You can change the kernel as needed (e.g., 'rbf' for radial basis function)
svm.fit(X_train, y_train)
svm_predictions = svm.predict(X_test)
svm_accuracy = accuracy_score(y_test, svm_predictions)

print("SVM Accuracy:", svm_accuracy)

Logistic Regression Accuracy: 0.6060606060606061
SVM Accuracy: 0.6060606060606061


#### 1 year offset results

Logistic Regression Accuracy: 0.631578947368421

SVM Accuracy: 0.6842105263157895

#### 2 year offset results

Logistic Regression Accuracy: 0.6060606060606061

SVM Accuracy: 0.6060606060606061

### Testing all attributes with SVM

In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split # Import train_test_split function

# Assuming X_train, y_train, X_test, y_test are your training and testing data
# X_train and X_test should be your feature matrices, and y_train and y_test should be your target labels (0 or 1)
features=['averageWinRate', 'averagePoints','averageRebounds','averageAssists', 'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio', 'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']
target = 'playoff'
x = df[features]
y = df[target]

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=1) # 70% training and 30% test

# Logistic Regression
logreg = LogisticRegression()

logreg = LogisticRegression(max_iter=1000)
logreg.fit(X_train, y_train)
logreg_predictions = logreg.predict(X_test)
logreg_accuracy = accuracy_score(y_test, logreg_predictions)
print("Logistic Regression Accuracy:", logreg_accuracy)

# Support Vector Machine
svm = SVC(kernel='linear')  # You can change the kernel as needed (e.g., 'rbf' for radial basis function)
svm.fit(X_train, y_train)
svm_predictions = svm.predict(X_test)
svm_accuracy = accuracy_score(y_test, svm_predictions)

print("SVM Accuracy:", svm_accuracy)

Logistic Regression Accuracy: 0.6363636363636364
SVM Accuracy: 0.6060606060606061


#### 1 year offset results

Logistic Regression Accuracy: 0.6842105263157895

SVM Accuracy: 0.6842105263157895

#### 2 year offset results

Logistic Regression Accuracy: 0.6363636363636364

SVM Accuracy: 0.6060606060606061

### Fine tuning and testing

#### Fine tuning on random split

In [3]:
from itertools import combinations
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import pandas as pd

In [16]:
features = ['averageWinRate', 'averagePoints', 'averageRebounds', 'averageAssists', 
            'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio', 
            'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']

target = 'playoff'  
X = df[features]
y = df[target]

# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

best_accuracy = 0
best_feature_combination = None

# Iterate through all possible feature combinations
for i in tqdm(range(1, len(features) + 1)):
    for subset in combinations(features, i):
        selected_features = list(subset)
        X_train_subset = X_train[selected_features]
        X_test_subset = X_test[selected_features]
        
        # Define hyperparameters and their possible values to search
        param_grid = {
            'C': [0.1, 1, 10],
            'penalty': ['l2']
        }
        
        # Create a Logistic Regression model
        model = LogisticRegression(solver='liblinear')
        
        # Perform grid search with cross-validation
        grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5)
        grid_search.fit(X_train_subset, y_train)
        
        # Get the best parameters and the best model
        best_params = grid_search.best_params_
        best_model = grid_search.best_estimator_
        
        # Evaluate the model on the test set
        accuracy = best_model.score(X_test_subset, y_test)
        
        # Check if this combination is the best so far
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_feature_combination = selected_features
            best_hyperparameters = best_params

print("Best Feature Combination:", best_feature_combination)
print("Best Hyperparameters:", best_hyperparameters)
print("Best Model Accuracy:", best_accuracy)


Best Feature Combination: ['averageWinRate', 'averagePoints', 'averageTurnovers']
Best Hyperparameters: {'C': 10, 'penalty': 'l2'}
Best Model Accuracy: 0.7575757575757576


#### 1 year offset results

Best Feature Combination: ['averageWinRate', 'averageRebounds', 'averageAssists', 'averageTurnovers', 'averageFGRatio', 'coachWinRate']

Best Hyperparameters: {'C': 1, 'penalty': 'l2'}

Best Model Accuracy: 0.8157894736842105

#### 2 year offset results

Best Feature Combination: ['averageWinRate', 'averagePoints', 'averageTurnovers']

Best Hyperparameters: {'C': 10, 'penalty': 'l2'}

Best Model Accuracy: 0.7575757575757576



### Fine tuning on split by years (train on oldest)

In [19]:
features = ['year', 'averageWinRate', 'averagePoints', 'averageRebounds', 'averageAssists', 
            'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio', 
            'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']

target = 'playoff'  
X = df[features]
y = df[target]

#Split dataset in such way that we use latest years for training and older years for testing
X_train=df[df["year"]<9][features]
X_test=df[df["year"]>=9][features]
y_train=df[df["year"]<9]["playoff"]
y_test=df[df["year"]>=9]["playoff"]

# Get percentage of data in each dataset
print("X_train: ", round((len(X_train) / (len(X_train) + len(X_test))) * 100, 2), "%")
print("X_test: ", round((len(X_test) / (len(X_train) + len(X_test))) * 100, 2), "%")
print("y_train: ", round((len(y_train) / (len(y_train) + len(y_test))) * 100, 2), "%")
print("y_test: ", round((len(y_test) / (len(y_train) + len(y_test))) * 100, 2), "%")


best_accuracy = 0
best_feature_combination = None

# Iterate through all possible feature combinations
for i in tqdm(range(1, len(features) + 1)):
    for subset in combinations(features, i):
        selected_features = list(subset)
        X_train_subset = X_train[selected_features]
        X_test_subset = X_test[selected_features]
        
        # Define hyperparameters and their possible values to search
        param_grid = {
            'C': [0.1, 1, 10],
            'penalty': ['l2']
        }
        
        # Create a Logistic Regression model
        model = LogisticRegression(solver='liblinear')
        
        # Perform grid search with cross-validation
        grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5)
        grid_search.fit(X_train_subset, y_train)
        
        # Get the best parameters and the best model
        best_params = grid_search.best_params_
        best_model = grid_search.best_estimator_
        
        # Evaluate the model on the test set
        accuracy = best_model.score(X_test_subset, y_test)
        
        # Check if this combination is the best so far
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_feature_combination = selected_features
            best_hyperparameters = best_params

print("Best Feature Combination:", best_feature_combination)
print("Best Hyperparameters:", best_hyperparameters)
print("Best Model Accuracy:", best_accuracy)


X_train:  75.45 %
X_test:  24.55 %
y_train:  75.45 %
y_test:  24.55 %


  0%|          | 0/13 [00:00<?, ?it/s]

100%|██████████| 13/13 [05:16<00:00, 24.37s/it]

Best Feature Combination: ['averageBlocks', 'averageFGRatio', 'averageFTRatio']
Best Hyperparameters: {'C': 10, 'penalty': 'l2'}
Best Model Accuracy: 0.7777777777777778


#### 1 year offset results

Best Feature Combination: ['averageTurnovers', 'averageFTRatio', 'coachWinRate', 'numberOfAwardedPlayers']

Best Hyperparameters: {'C': 10, 'penalty': 'l2'}

Best Model Accuracy: 0.7777777777777778

#### 2 year offset results

Best Feature Combination: ['averageBlocks', 'averageFGRatio', 'averageFTRatio']

Best Hyperparameters: {'C': 10, 'penalty': 'l2'}

Best Model Accuracy: 0.7777777777777778


### Fine tuning on split by years (train on latest)

In [20]:
features = ['averageWinRate', 'averagePoints', 'averageRebounds', 'averageAssists', 
            'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio', 
            'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']

target = 'playoff'  
X = df[features]
y = df[target]

#Split dataset in such way that we use latest years for training and older years for testing
X_train=df[df["year"]>3][features]
X_test=df[df["year"]<=3][features]
y_train=df[df["year"]>3]["playoff"]
y_test=df[df["year"]<=3]["playoff"]

# Get percentage of data in each dataset
print("X_train: ", round((len(X_train) / (len(X_train) + len(X_test))) * 100, 2), "%")
print("X_test: ", round((len(X_test) / (len(X_train) + len(X_test))) * 100, 2), "%")
print("y_train: ", round((len(y_train) / (len(y_train) + len(y_test))) * 100, 2), "%")
print("y_test: ", round((len(y_test) / (len(y_train) + len(y_test))) * 100, 2), "%")

best_accuracy = 0
best_feature_combination = None

# Iterate through all possible feature combinations
for i in tqdm(range(1, len(features) + 1)):
    for subset in combinations(features, i):
        selected_features = list(subset)
        X_train_subset = X_train[selected_features]
        X_test_subset = X_test[selected_features]
        
        # Define hyperparameters and their possible values to search
        param_grid = {
            'C': [0.1, 1, 10],
            'penalty': ['l2']
        }
        
        # Create a Logistic Regression model
        model = LogisticRegression(solver='liblinear')
        
        # Perform grid search with cross-validation
        grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5)
        grid_search.fit(X_train_subset, y_train)
        
        # Get the best parameters and the best model
        best_params = grid_search.best_params_
        best_model = grid_search.best_estimator_
        
        # Evaluate the model on the test set
        accuracy = best_model.score(X_test_subset, y_test)
        
        # Check if this combination is the best so far
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_feature_combination = selected_features
            best_hyperparameters = best_params

print("Best Feature Combination:", best_feature_combination)
print("Best Hyperparameters:", best_hyperparameters)
print("Best Model Accuracy:", best_accuracy)


X_train:  85.45 %
X_test:  14.55 %
y_train:  85.45 %
y_test:  14.55 %


100%|██████████| 12/12 [02:32<00:00, 12.70s/it]

Best Feature Combination: ['averagePoints', 'averageRebounds', 'averageTurnovers', 'numberOfAwardedPlayers']
Best Hyperparameters: {'C': 0.1, 'penalty': 'l2'}
Best Model Accuracy: 0.75


### Different approach to hyperparam tuning

Saves results to file for later use.

In [ ]:
# Model training...
from itertools import combinations, product
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from tqdm import tqdm

# Assume df is your DataFrame
# Define features and target
features = ['averageWinRate', 'averagePoints', 'averageRebounds', 'averageAssists',
            'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio',
            'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']
target = 'playoff'

# Split data into training and test sets
X_train = df[df["year"] < 7][features + ['year']]
X_test = df[df["year"] >= 7][features + ['year']]
y_train = df[df["year"] < 7]["playoff"]
y_test = df[df["year"] >= 7]["playoff"]

df_year_only_train = X_train["year"]
df_year_only_test = X_test["year"]

best_accuracy_sum = 0
best_feature_combination = None
best_hyperparameters = None

# Define hyperparameters grid
param_grid = {
    'C': [0.1, 1, 10],
    'penalty': ['l2'],
    'solver': ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga'],
    'tol': [1e-4, 1e-3, 1e-2],
    'max_iter': [10000]
}

# Generate all combinations of hyperparameters
all_params = [dict(zip(param_grid, v)) for v in product(*param_grid.values())]

# Iterate through all possible feature combinations with 2 to 6 features
results = []
for subset in tqdm(combinations(features, 4), desc="Feature Combinations"):
    selected_features = list(subset) + ['year']

    X_train_subset = X_train[selected_features]
    X_test_subset = X_test[selected_features]

    for params in all_params:
        model = LogisticRegression(**params)
        model.fit(X_train_subset, y_train)

        accuracy_test = accuracy_score(y_test, model.predict(X_test_subset))
        accuracy_train = accuracy_score(y_train, model.predict(X_train_subset))

        if accuracy_test > 0.75:
            results.append({
                'acc_train': accuracy_train,
                'acc_test': accuracy_test,
                'best_feature_combination': str(selected_features),
                'best_hyperparameters': str(params)
            })

# Convert results to DataFrame
results_df = pd.DataFrame(results)

df_cd_json = results_df.to_json(orient="records")
new_file = open("./datasets/generated/logreg_hyper_param_tuning_2.json", "w")
new_file.writelines(df_cd_json)
new_file.close()

#### 1 year offset results

Best Feature Combination: ['averageAssists', 'averageFGRatio', 'numberOfAwardedPlayers']

Best Hyperparameters: {'C': 10, 'penalty': 'l2'}

Best Model Accuracy: 0.8125

#### 2 year offset results

Best Feature Combination: ['averagePoints', 'averageRebounds', 'averageTurnovers', 'numberOfAwardedPlayers']

Best Hyperparameters: {'C': 0.1, 'penalty': 'l2'}

Best Model Accuracy: 0.75

## KNN 


In [4]:
from itertools import combinations, product
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from tqdm import tqdm

features = ['averageWinRate', 'averagePoints', 'averageRebounds', 'averageAssists',
            'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio',
            'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers',]
target = 'playoff'

#split data
X_train = df[df["year"] < 7][features + ['year']]
X_test = df[df["year"] >= 7][features + ['year']]
y_train = df[df["year"] < 7]["playoff"]
y_test = df[df["year"] >= 7]["playoff"]

grid_params = { 'n_neighbors' : [5,7,9,11,13,15],
               'weights' : ['uniform','distance'],
               'metric' : ['minkowski','euclidean','manhattan']}

#generate all combinations of the parameters
all_params = [dict(zip(grid_params, v)) for v in product(*grid_params.values())]


results = pd.DataFrame(columns=['acc_train', 'acc_test', 'best_feature_combination', 'best_hyperparameters'])

for num_features in range(2, 7): 
    for subset in tqdm(combinations(features, num_features), desc=f"Feature Combinations {num_features}"):
        selected_features = list(subset)
        #this way year is always included
        selected_features.append('year')

        X_train_subset = X_train[selected_features]
        X_test_subset = X_test[selected_features]

        for params in all_params:

            model = KNeighborsClassifier(**params)
            model.fit(X_train_subset, y_train)

            #evaluate the model
            y_pred_test = model.predict(X_test_subset)
            accuracy_test = accuracy_score(y_test, y_pred_test)
            y_pred_train = model.predict(X_train_subset)
            accuracy_train = accuracy_score(y_train, y_pred_train)

            if accuracy_test > 0.65:
                temp_row = pd.DataFrame([{
                    'acc_train': accuracy_train,
                    'acc_test': accuracy_test,
                    'best_feature_combination': str(selected_features),
                    'best_hyperparameters': str(params)
                }])
                results = pd.concat([results, temp_row])

# Save the results to a JSON file
df_cd_json = results.to_json(orient="records")
new_file = open("./datasets/generated/knn_hyper_param_tuning.json", "w")
new_file.writelines(df_cd_json)
new_file.close()

results.head()



Feature Combinations 2: 0it [00:00, ?it/s]


KeyError: "['year'] not in index"

## SVM

In [17]:
from itertools import combinations, product
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score
from tqdm import tqdm
from sklearn.svm import SVC

# Assume df is your DataFrame
# Define features and target
features = ['averageWinRate', 'averagePoints', 'averageRebounds', 'averageAssists',
            'averageSteals', 'averageBlocks', 'averageTurnovers', 'averageFGRatio',
            'averageFTRatio', 'averageThreeRatio', 'coachWinRate', 'numberOfAwardedPlayers']
target = 'playoff'

# Split data into training and test sets
X_train = df[df["year"] < 7][features + ['year']]
X_test = df[df["year"] >= 7][features + ['year']]
y_train = df[df["year"] < 7]["playoff"]
y_test = df[df["year"] >= 7]["playoff"]

df_year_only_train = X_train["year"]
df_year_only_test = X_test["year"]

model = SVC()
param_grid = {
    'C': [0.1, 1, 10, 100],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': ['scale', 'auto', 0.1, 1, 10]
}
 
best_accuracy_sum = 0
best_feature_combination = None
best_hyperparameters = None

# Generate all combinations of hyperparameters
all_params = [dict(zip(param_grid, v)) for v in product(*param_grid.values())]

# Iterate through all possible feature combinations with 2 to 6 features
results = pd.DataFrame(columns=['acc_train', 'acc_test', 'best_feature_combination', 'best_hyperparameters'])

# Always include 'year

for subset in tqdm(combinations(features, 4), desc="Feature Combinations"):
    selected_features = list(subset) + ['year']

    X_train_subset = X_train[selected_features]
    X_test_subset = X_test[selected_features]

    for params in all_params:
       
        grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='accuracy')
        grid_search.fit(X_train_subset, y_train)
        
        best_model = grid_search.best_estimator_
        test_accuracy = best_model.score(X_test, y_test)
        
        accuracy_test = best_model.score(y_test, model.predict(X_test_subset))
        accuracy_train = best_model.score(y_train, model.predict(X_train_subset))

        results.append({
            'acc_train': accuracy_train,
            'acc_test': accuracy_test,
            'best_feature_combination': str(selected_features),
            'best_hyperparameters': str(best_model)
        }, ignore_index=True)

# Convert results to DataFrame
results_df = pd.DataFrame(results)

results_df

df_cd_json = results_df.to_json(orient="records")
new_file = open("./datasets/generated/svm_hyper_param_tuning.json", "w")
new_file.writelines(df_cd_json)
new_file.close()
 

Feature Combinations: 0it [00:00, ?it/s]

## Conclusion

| Split | Year offset | Best feature combination | Best hyperparameters | Best model accuracy |
| ----- | ----------- | ------------------------ | -------------------- | ------------------- |
| Random split (70% train, 30% test) | 1 | 'averageWinRate', 'averageRebounds', 'averageAssists', 'averageTurnovers', 'averageFGRatio', 'coachWinRate' | {'C': 1, 'penalty': 'l2'} | 81.58% |
| Split by years, train on oldest 7 years | 1 | 'averageSteals', 'numberOfAwardedPlayers' | {'C': 10, 'penalty': 'l2'} | 72.5% |
| Split by years, train on latest 7 years | 1 | 'averageAssists', 'averageFGRatio', 'numberOfAwardedPlayers' | {'C': 10, 'penalty': 'l2'} | 81.25% |


By comparing the models' accuracy, we can see that the fine tuned model with the random split has the greatest accuracy of 81.58%.